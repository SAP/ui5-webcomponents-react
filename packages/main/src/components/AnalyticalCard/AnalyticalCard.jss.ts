import { ThemingParameters, spacing } from '@ui5/webcomponents-react-base';

const styles = {
  card: {
    backgroundColor: ThemingParameters.sapTile_Background,
    border: `0.0625rem solid ${ThemingParameters.sapTile_BorderColor}`,
    boxShadow: ThemingParameters.sapContent_Shadow0,
    borderRadius: ThemingParameters.sapElement_BorderCornerRadius,
    textAlign: 'start',
    overflow: 'hidden',
    position: 'relative',
    fontFamily: ThemingParameters.sapFontFamily,
    boxSizing: 'border-box',
    width: '20rem'
  },
  content: spacing.sapUiContentPadding
};

export default styles;
