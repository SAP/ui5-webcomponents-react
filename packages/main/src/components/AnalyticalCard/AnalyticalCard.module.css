.content {
  display: flex;
  flex: 1 1 auto;
  overflow: hidden;
  position: relative;
}
