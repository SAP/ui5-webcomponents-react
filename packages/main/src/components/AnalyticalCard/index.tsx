import React, { forwardRef, ReactNode, Ref } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { Card, CardDomRef } from '../../webComponents';
import styles from './AnalyticalCard.jss';

export interface AnalyticalCardPropTypes extends CommonProps {
  /**
   * The Card header Component. Using the `AnalyticalCardHeader` is recommended.
   */
  header?: ReactNode;
  /**
   * The content of the `AnalyticalCard`.
   */
  children: ReactNode | ReactNode[];
}

const useStyles = createUseStyles(styles, { name: 'AnalyticalCard' });
/**
 * The `AnalyticalCard` is mainly used for data visualization. It consists of two areas â€“ a header area and a chart area with a visual representation of the data.<br />
 */
const AnalyticalCard = forwardRef((props: AnalyticalCardPropTypes, ref: Ref<CardDomRef>) => {
  const { children, header, ...rest } = props;
  const classes = useStyles();

  return (
    <Card ref={ref} {...rest}>
      {header}
      <div className={classes.content} aria-role="group">
        {children}
      </div>
    </Card>
  );
});

AnalyticalCard.displayName = 'AnalyticalCard';

export { AnalyticalCard };
