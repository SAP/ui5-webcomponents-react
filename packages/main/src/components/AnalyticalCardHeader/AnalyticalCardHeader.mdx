import { ArgTypesWithNote, ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import SubcomponentsSection from '@sb/docs/SubcomponentsSection.md?raw';
import { Canvas, Description, Markdown, Meta } from '@storybook/blocks';
import '@ui5/webcomponents-icons/dist/person-placeholder.js';
import '@ui5/webcomponents/dist/Link.js';
import { NumericSideIndicator } from '../NumericSideIndicator';
import * as ComponentStories from './AnalyticalCardHeader.stories';

<Meta of={ComponentStories} />

<DocsHeader subComponents={['NumericSideIndicator']} />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<br />
<br />

<Markdown>{SubcomponentsSection}</Markdown>

## NumericSideIndicator

<Description of={NumericSideIndicator} />

<ArgTypesWithNote of={NumericSideIndicator} />

<Footer />
