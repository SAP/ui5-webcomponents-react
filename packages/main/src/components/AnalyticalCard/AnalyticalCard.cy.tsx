import { getRGBColor } from '@ui5/webcomponents-base/dist/util/ColorConversion';
import { ThemingParameters } from '@ui5/webcomponents-react-base';
import { DeviationIndicator, ValueColor } from '../../enums';
import { AnalyticalCardHeader, AnalyticalCardHeaderPropTypes } from '../AnalyticalCardHeader';
import { NumericSideIndicator } from '../NumericSideIndicator';
import { Text } from '../Text';
import { AnalyticalCard, AnalyticalCardPropTypes } from './index';

const TestComp = (props: Omit<AnalyticalCardPropTypes, 'children'>) => {
  return (
    <AnalyticalCard {...props}>
      <Text>Content</Text>
    </AnalyticalCard>
  );
};

describe('AnalyticalCard', () => {
  it('Render with default AnalyticalCardHeader', () => {
    cy.mount(<TestComp header={<AnalyticalCardHeader titleText="Header Title" />} />);
    cy.findByText('Content').should('be.visible');
    cy.findByText('Header Title').should('be.visible');
  });

  it('Render with custom header', () => {
    cy.mount(<TestComp header={<div>Header Title</div>} />);
    cy.findByText('Content').should('be.visible');
    cy.findByText('Header Title').should('be.visible');
  });

  [
    [ValueColor.Error, ThemingParameters.sapNegativeTextColor],
    [ValueColor.Critical, ThemingParameters.sapCriticalTextColor],
    [ValueColor.Good, ThemingParameters.sapPositiveTextColor],
    [ValueColor.Neutral, ThemingParameters.sapNeutralTextColor]
  ].forEach(([color, cssVar]: [AnalyticalCardHeaderPropTypes['state'], string]) => {
    it(`Header: ValueColor: ${color}`, () => {
      const cssVarValue = getComputedStyle(document.documentElement).getPropertyValue(cssVar.match(/(--)[^)]+/)[0]);
      cy.mount(
        <TestComp header={<AnalyticalCardHeader trend={DeviationIndicator.Down} value="123" state={color} />} />
      );
      const rgbVal = getRGBColor(cssVarValue);
      const rgbValString = `rgb(${rgbVal.r}, ${rgbVal.g}, ${rgbVal.b})`;
      cy.findByText('123').parent().should('have.css', 'color', rgbValString);
    });
  });

  it('Header: all string props', () => {
    cy.mount(
      <TestComp
        header={
          <AnalyticalCardHeader
            titleText="Title"
            subtitleText="Subtitle"
            value="Value"
            unitOfMeasurement="Unit"
            description="Description"
          >
            <NumericSideIndicator titleText="Target" number={100} />
            <NumericSideIndicator titleText="Deviation" number={50} state={ValueColor.Error} />
          </AnalyticalCardHeader>
        }
      />
    );
    cy.findByText('Title').should('be.visible');
    cy.findByText('Subtitle').should('be.visible');
    cy.findByText('Value').should('be.visible');
    cy.findByText('Unit').should('be.visible');
    cy.findByText('Target').should('be.visible');
    cy.findByText('Deviation').should('be.visible');
  });

  it('Header: with arrow indicator', () => {
    cy.mount(<TestComp header={<AnalyticalCardHeader scale="EUR" trend={DeviationIndicator.Down} />} />);
    cy.get('[ui5-icon]').should('have.attr', 'name', 'down');
    cy.mount(<TestComp header={<AnalyticalCardHeader scale="EUR" trend={DeviationIndicator.Up} />} />);
    cy.get('[ui5-icon]').should('have.attr', 'name', 'up');
  });

  // todo custom props test for header and card
});
