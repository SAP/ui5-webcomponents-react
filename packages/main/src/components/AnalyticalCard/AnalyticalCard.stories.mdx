import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { DocsHeader } from '@docs/DocsHeader';
import { Footer } from '@docs/Footer';
import '@ui5/webcomponents-icons/dist/person-placeholder.js';
import { AnalyticalCard } from './index';
import { AnalyticalCardHeader } from '../AnalyticalCardHeader';
import { DeviationIndicator } from '../../enums/DeviationIndicator';
import { ValueState } from '../../enums/ValueState';
import { LineChart } from '@ui5/webcomponents-react-charts';

<Meta
  title="Data Display / AnalyticalCard"
  component={AnalyticalCard}
  subcomponents={{ AnalyticalCardHeader }}
  argTypes={{
    header: {
      control: { disable: true }
    },
    children: {
      control: { disable: true }
    }
  }}
/>

<DocsHeader />

**Note:** The `AnalyticalCard` was mainly made to display KPIs and complex data. If you want to implement a simple card, the `Card` component probably is more suitable.

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      const simpleDataSet = [
        {
          name: 'January',
          users: 76
        },
        {
          name: 'February',
          users: 230
        },
        {
          name: 'March',
          users: 240
        },
        {
          name: 'April',
          users: 280
        },
        {
          name: 'May',
          users: 100
        }
      ];
      return (
        <AnalyticalCard
          header={
            <AnalyticalCardHeader
              titleText="Title"
              subtitleText="Sub Title"
              arrowIndicator={DeviationIndicator.Down}
              indicatorState={ValueState.Success}
              value="Value"
              valueState={ValueState.Success}
              unit="Unit"
              target="Target"
              deviation="Deviation"
              showIndicator
              description="Description"
              counter="Counter"
              counterState={ValueState.Success}
              currency="EUR"
            />
          }
        >
          <LineChart
            noLegend
            dimensions={[{ accessor: 'name' }]}
            measures={[{ accessor: 'users', formatter: (val) => `${val}k` }]}
            dataset={simpleDataSet}
          />
        </AnalyticalCard>
      );
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="Default" />

<br />
<br />

# More Examples

<br />

## Customizable AnalyticalCardHeader

<Canvas>
  <Story
    name="Customizable header"
    args={{
      titleText: 'Project Cost',
      subtitleText: 'All active Projects',
      arrowIndicator: DeviationIndicator.Down,
      indicatorState: ValueState.Success,
      value: '100',
      valueState: ValueState.Success,
      unit: 'M',
      target: '120M',
      deviation: '-16.67%',
      showIndicator: true,
      description: 'Current Year',
      counter: '+',
      counterState: ValueState.Success,
      currency: 'USD'
    }}
    argTypes={{
      titleText: { description: 'Defines the title of the `AnalyticalCardHeader`.' },
      subtitleText: { description: 'Defines the sub title of the `AnalyticalCardHeader`.' },
      value: { description: 'Defines the value of the `AnalyticalCardHeader`.' },
      unit: { description: 'Defines the unit displayed next to the value of the `AnalyticalCardHeader`.' },
      target: { description: 'Defines the target value.' },
      deviation: { description: 'Defines the deviation value.' },
      showIndicator: { description: 'Defines whether the deviation indicator should be displayed.' },
      description: { description: 'Defines the description below the value of the `AnalyticalCardHeader`.' },
      counter: { description: 'Defines the counter in the upper right corner of the `AnalyticalCardHeader`.' },
      currency: { description: 'Defines the currency.' },
      onClick: { description: 'Fired when the `AnalyticalCardHeader` header is clicked.' },
      header: { table: { disable: true } },
      children: { table: { disable: true } },
      ref: { table: { disable: true } }
    }}
  >
    {(args) => {
      const simpleDataSet = [
        {
          name: 'January',
          users: 76
        },
        {
          name: 'February',
          users: 130
        },
        {
          name: 'March',
          users: 140
        },
        {
          name: 'April',
          users: 180
        },
        {
          name: 'May',
          users: 100
        }
      ];
      return (
        <AnalyticalCard
          header={
            <AnalyticalCardHeader
              titleText={args.titleText}
              subtitleText={args.subtitleText}
              arrowIndicator={args.arrowIndicator}
              indicatorState={args.indicatorState}
              value={args.value}
              valueState={args.valueState}
              unit={args.unit}
              target={args.target}
              deviation={args.deviation}
              onClick={args.onClick}
              showIndicator={args.showIndicator}
              description={args.description}
              counter={args.counter}
              counterState={args.counterState}
              currency={args.currency}
            />
          }
        >
          <LineChart
            noLegend
            dimensions={[{ accessor: 'name' }]}
            measures={[{ accessor: 'users', formatter: (val) => `${val}m` }]}
            dataset={simpleDataSet}
          />
        </AnalyticalCard>
      );
    }}
  </Story>
</Canvas>

<ArgsTable story="Customizable header" />

<Footer />
