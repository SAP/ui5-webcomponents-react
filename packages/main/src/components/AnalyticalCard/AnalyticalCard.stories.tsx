import type { Meta, StoryObj } from '@storybook/react';
import { LineChart } from '@ui5/webcomponents-react-charts';
import { DeviationIndicator, ValueColor } from '../../enums';
import { AnalyticalCardHeader } from '../AnalyticalCardHeader';
import { NumericSideIndicator } from '../NumericSideIndicator';
import { AnalyticalCard } from './index.js';

const simpleDataSet = [
  {
    name: 'January',
    users: 76
  },
  {
    name: 'February',
    users: 230
  },
  {
    name: 'March',
    users: 240
  },
  {
    name: 'April',
    users: 280
  },
  {
    name: 'May',
    users: 100
  }
];

const meta = {
  title: 'Data Display / AnalyticalCard',
  component: AnalyticalCard,
  argTypes: {
    header: {
      control: { disable: true }
    },
    children: {
      control: { disable: true }
    }
  },
  args: {
    children: (
      <LineChart
        className="chromatic-ignore"
        noLegend
        dimensions={[{ accessor: 'name' }]}
        measures={[{ accessor: 'users', formatter: (val) => `${val}k` }]}
        dataset={simpleDataSet}
      />
    ),
    header: (
      <AnalyticalCardHeader
        titleText="Project Cloud Transformation"
        subtitleText="Revenue"
        unitOfMeasurement="EUR"
        trend={DeviationIndicator.Down}
        value="65.34"
        state={ValueColor.Error}
        scale="K"
        description="Q1, 2018"
      >
        <NumericSideIndicator titleText="Target" number="100" unit="k" />
        <NumericSideIndicator titleText="Deviation" number="34.7" unit="%" state={ValueColor.Critical} />
      </AnalyticalCardHeader>
    )
  }
} satisfies Meta<typeof AnalyticalCard>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
