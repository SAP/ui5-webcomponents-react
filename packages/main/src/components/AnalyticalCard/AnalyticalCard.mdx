import { ArgTypesWithNote, ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import SubcomponentsSection from '@sb/docs/SubcomponentsSection.md?raw';
import { Canvas, Description, Markdown, Meta } from '@storybook/blocks';
import '@ui5/webcomponents-icons/dist/person-placeholder.js';
import { AnalyticalCardHeader } from '../AnalyticalCardHeader';
import { NumericSideIndicator } from '../NumericSideIndicator';
import * as ComponentStories from './AnalyticalCard.stories';

<Meta of={ComponentStories} subcomponents={{ AnalyticalCardHeader, NumericSideIndicator }} />

<DocsHeader />

**Note:** The `AnalyticalCard` was mainly made to display KPIs and complex data. If you want to implement a simple card, the `Card` component probably is more suitable.

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<br />
<br />

<Markdown>{SubcomponentsSection}</Markdown>

## AnalyticalCardHeader

<Description of={AnalyticalCardHeader} />

<ArgTypesWithNote of={AnalyticalCardHeader} />

## NumericSideIndicator

<Description of={NumericSideIndicator} />

<ArgTypesWithNote of={NumericSideIndicator} />

<Footer />
