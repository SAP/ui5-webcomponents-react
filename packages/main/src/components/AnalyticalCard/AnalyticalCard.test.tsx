import { render, screen } from '@shared/tests';
import { createCustomPropsTest } from '@shared/tests/utils';
import { AnalyticalCard } from '@ui5/webcomponents-react/dist/AnalyticalCard';
import { AnalyticalCardHeader } from '@ui5/webcomponents-react/dist/AnalyticalCardHeader';
import { Text } from '@ui5/webcomponents-react/dist/Text';
import React from 'react';

const Header = <AnalyticalCardHeader titleText="Header Title" />;

describe('Analytical Card', () => {
  test('Render without Crashing', () => {
    const { asFragment } = render(
      <AnalyticalCard>
        <Text>I'm a content!</Text>
      </AnalyticalCard>
    );
    expect(screen.getByText("I'm a content!")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('with custom header', () => {
    const { asFragment } = render(
      <AnalyticalCard header={<div>I'm a header!</div>}>
        <Text>I'm a content!</Text>
      </AnalyticalCard>
    );
    expect(screen.getByText("I'm a content!")).toBeInTheDocument();
    expect(screen.getByText("I'm a header!")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('with AnalyticalCardHeader', () => {
    const { asFragment } = render(
      <AnalyticalCard header={Header}>
        <Text>I'm a content!</Text>
      </AnalyticalCard>
    );
    expect(screen.getByText('Header Title')).toBeInTheDocument();
    expect(screen.getByText("I'm a content!")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  createCustomPropsTest(AnalyticalCard);
});
