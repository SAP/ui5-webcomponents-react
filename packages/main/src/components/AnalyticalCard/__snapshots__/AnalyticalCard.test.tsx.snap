// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytical Card Render without Crashing 1`] = `
<DocumentFragment>
  <ui5-card
    ui5-card=""
  >
    <div
      class="AnalyticalCard-content"
      role="group"
    >
      <span
        class="Text-text"
      >
        I'm a content!
      </span>
    </div>
  </ui5-card>
</DocumentFragment>
`;

exports[`Analytical Card with AnalyticalCardHeader 1`] = `
<DocumentFragment>
  <ui5-card
    ui5-card=""
  >
    <div
      aria-labelledby="1337-title 1337-mainIndicator"
      aria-roledescription="Card Header"
      class="AnalyticalCardHeader-cardHeader"
      data-sap-ui-fastnavgroup="true"
      id="1337"
      role="heading"
      slot="header"
      tabindex="0"
    >
      <div>
        <div
          class="AnalyticalCardHeader-headerTitles"
        >
          <div
            class="AnalyticalCardHeader-headerFirstLine"
          >
            <span
              aria-level="3"
              class="AnalyticalCardHeader-headerText"
              id="1337-title"
              role="heading"
            >
              Header Title
            </span>
          </div>
        </div>
        <div
          class="AnalyticalCardHeader-kpiContent"
        >
          <div
            aria-label="undefinedundefined
"
            class="AnalyticalCardHeader-mainIndicator"
            id="1337-mainIndicator"
            role="img"
          >
            <span
              class="AnalyticalCardHeader-value"
            />
            <div
              class="AnalyticalCardHeader-indicatorAndUnit"
            >
              <div
                class="AnalyticalCardHeader-unit"
              />
            </div>
          </div>
          <div
            class="AnalyticalCardHeader-indicatorGap"
          />
          <div
            class="AnalyticalCardHeader-sideIndicators"
          />
        </div>
      </div>
    </div>
    <div
      class="AnalyticalCard-content"
      role="group"
    >
      <span
        class="Text-text"
      >
        I'm a content!
      </span>
    </div>
  </ui5-card>
</DocumentFragment>
`;

exports[`Analytical Card with custom header 1`] = `
<DocumentFragment>
  <ui5-card
    ui5-card=""
  >
    <div>
      I'm a header!
    </div>
    <div
      class="AnalyticalCard-content"
      role="group"
    >
      <span
        class="Text-text"
      >
        I'm a content!
      </span>
    </div>
  </ui5-card>
</DocumentFragment>
`;
